### PR TITLE
feat: Add context usage to libmirror push operations

### PR DESCRIPTION
**Pull request by mvasl**
Added `context.Context` usage to push-related functions of libmirror

### CODE REVIEW

File: pkg/libmirror/util/retry/retry.go [hunk 1]
```diff
@@ -1,30 +1,39 @@
...
 	var lastErr error
 	for restarts < task.MaxRetries() {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
 		if restarts > 0 {
 			interval := task.Interval(restarts)
 			logger.InfoF("%s failed, next retry in %v", name, interval)
 			time.Sleep(interval)
```
Comment (vadimartynov) [inline pkg/libmirror/util/retry/retry.go:32]: The function sleep might hang. It's better to implement it using select:

```
			select {
			case <-time.After(interval):
				// Pause completed, proceed with next attempt
			case <-ctx.Done():
				return fmt.Errorf("%q: task cancelled during retry wait, last error: %w", name, ctx.Err())
			}
```

Comment (mvasl) [reply pkg/libmirror/util/retry/retry.go]: Thanks

**Comment by vadimartynov** [comment]
Looks fire 🔥

File: pkg/libmirror/util/auth/auth.go [hunk 2]
```diff
@@ -32,14 +31,21 @@ import (
...
+	insecure, skipVerifyTLS bool,
+) error {
 	nameOpts, remoteOpts := MakeRemoteRegistryRequestOptions(authProvider, insecure, skipVerifyTLS)
 	ref, err := name.ParseReference(imageTag, nameOpts...)
 	if err != nil {
 		return fmt.Errorf("Parse registry address: %w", err)
 	}
 
-	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
-	defer cancel()
```
[the remark this answers is not in the document]
Comment (vadimartynov) [reply pkg/libmirror/util/auth/auth.go:42]: As they say, up to you :)
I just noticed. A timeout was previously used in the context, but now it has been removed. The main thing is to ensure it doesn’t affect the functionality of other parts of the system. It doesn’t impact the current task.

File: pkg/libmirror/layouts/push.go [hunk 3]
```diff
@@ -84,59 +109,61 @@ func PushLayoutToRepo(
...
-			parallel.ForEach(manifestSet, pushImage(logger, registryRepo, index, imagesCount, refOpts, remoteOpts))
+			errMu := &sync.Mutex{}
+			merr := &multierror.Error{}
+			parallel.ForEach(manifestSet, func(item v1.Descriptor, i int) {
+				if err = pushImage(ctx, registryRepo, index, item, refOpts, remoteOpts); err != nil {
+					errMu.Lock()
+					defer errMu.Unlock()
+					merr = multierror.Append(merr, err)
+				}
+			})
```
[the remark this answers is not in the document]
Comment (vadimartynov) [reply pkg/libmirror/layouts/push.go:120]: If we look at it from this perspective, then you're right.

File: pkg/libmirror/util/auth/auth.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mvasl) [reply pkg/libmirror/util/auth/auth.go:42]: Well, timeout was moved to the caller, not removed entirely. `d8 mirror` might use longer timeout this way, to adapt to slower registries. `deckhouse-registry` might want to make it shorter. Previously this was not possible, I see it as a win.

File: pkg/libmirror/layouts/push.go [hunk 4]
```diff
@@ -84,59 +109,61 @@ func PushLayoutToRepo(
...
 
-			return nil
+			return merr.ErrorOrNil()
 		})
 		if err != nil {
 			return fmt.Errorf("Push batch of images: %w", err)
 		}
-		batchesCount += 1
```
[the remark this answers is not in the document]
Comment (mvasl) [reply pkg/libmirror/layouts/push.go]: Fixed

**Pull request merged**